### ISSUE TITLE
Moving into C++11 concurrency objects

### ISSUE THREAD

**Issue by David-Haim-zz** [rescrv/HyperLevelDB#19]
Hi there,
I proposed the same suggestion to the original levelDB maintainers but the suggestion was refused.

Currently, (Hyper)LevelDB uses per-OS concurrency primitives in order to parallelize executation. by moving into C++11 concurrency primitives we can have much less code (no "xxx_port.cc" files) while maintaining the performance of the library and achieving cross platform compatible code. 

Also, we are using LevelDB in our Windows clients that have to be supported on XP as well. as it currently stands, the function "InitOnceExecuteOnce" is not available on XP. that means we need to change the port each time Visual Studio's Nuget package manager installs LevelDB. using C++11 will solve this problem as well.

what do you think of that idea?

**Comment by rescrv** [comment]
I think the C++ concurrency primitives do not solve any problem I have.  I know exactly the semantics provided by the atomics within leveldb (and the assembly instructions they translate to), whereas I'm unsure of what the C++ atomics will do in practice (because I haven't used them).

They also introduce a big problem:  portability.  C++11 is only portable to platforms that have a c++11 compiler, and many platforms do not have one, and those that do may not default to it.

In short, I don't think you'll have less code, but, in fact, more code.  You'll have everything you have now for the non-C++11 systems, and then a special "C++11 OS".  The performance may or may not be the same, and the cross-platform compatibility will be no better than what is there now.
